### DISCUSSION TITLE
Add fastcgi support to improve proxying support

### DISCUSSION THREAD

**Discussion by ghost** [verdaccio/verdaccio#2224] [Ideas]
Typically a proxied web service might use fastcgi.

Unix socks are typically used to provide this, and would allow a large installation to be more secure as you can provide services which are managed at a OS level rather than a networking level where services are exposed for everyone to access.

It allows the HTTP processor (proxy) to handle the web connections by offloading the HTTP header handling and allows the proxied service to be more efficient as a result as the two systems can scale independently.

Node is not the best web server under high concurrent loads - check any benchmark you like and you will see that anything JS is typically much much slower. 

https://stressgrid.com/blog/webserver_benchmark/

Anything that can be done to remove load from the service would be welcome.

**Comment by juanpicado** [comment]
I don't understand your request, maybe if you elaborate more on your idea and how fits this project. (I've moved it to discussion since there is not a clear action plan)

**Reply by flipkickmedia** [reply]
**What**
Well, typically when you proxy anything, you are attempting to move the load of load balancing, HTTP handling off the service to the proxy, otherwise why would you proxy?

Before fastcgi, everything was just proxied directly, so HTTP would be just sent as is to the service behind.  This meant that you were doing two lots of HTTP processing which actually in an HTTP session is the most inefficient method.  If you bring in HTTP2 you then end up with the proxied services being forwarded the connections from the proxy so you end up in a position where you can run out of TCP connection available.

In order to solve this and improve concurrent performance when scaling you might launch multiple services...and off load say SSL, HTTP(s) and authorisation handling into a separate system.

This system is known as fastcgi and allows the HTTP portion to be handled by the proxy. The service is then passed a fastcgi request which is typically smaller and driven over a binary protocol.  This leaves the connection management and HTTP to be handed by the proxy and means your service handling the connection us much more efficient and easier to manage.

Pretty much all stacks support fastcgi including node.js - and should be a drop in replacement https://www.npmjs.com/package/node-fastcgi

**Why**
If you want to deploy a large scale NPM repo, which can handle a lot of users, without the performance hit and complexities of securing that system, fastcgi solves a lot of the headaches.  HTTP to HTTP proxies are typically very slow and node.js is not a fast webserver.  Let Nginx/Apache handle that.
